### DISCUSSION TITLE
[QUESTION] Specifying ordering of tests

### DISCUSSION THREAD

**Discussion by ravy** [schemathesis/schemathesis#2364] [Q&A]
### question:
For swagger 2.0 docs, links are not supported. 
using dredd, able to order the apis using hooks and custom sort function 
to specify the exact order the apis need to be tested
using schemathesis,
using `--hypothesis-phases=explicit` with examples provided for all api, which is the test data
but unsure how to enforce order

### discussion:
currently see two ways around this:
1. programmatically introduce links in the schema after using a tool to convert docs from openapi 2.0 to 3.0
2. enforce order in schemathesis hooks

is there any schemathesis recipe/hook that can help me enforce order or openapi links is the way to go?

for example:
lets say we have 3 api on a resource named 'topic':
POST topics
DELETE topics/id
GET topics/id

and 3 api on a resource named 'subtopic' that exists under 'topic'
POST topic/id/subtopics
DELETE topic/id/subtopics/id2
GET topic/id/subtopics/id2

want to have a test case in this order exactly:
POST topics > GET topics > POST subtopics > GET subtopics > DELETE subtopics > DELETE topics

**Comment by Stranger6667** [comment]
Thanks for opening this! I'll get back to it a bit later, but meanwhile, I want to say that Schemathesis supports the `x-links` extension on Swagger 2.0 that has the same structure as `links` on Open API 3.x

**Comment by Stranger6667** [comment]
Schemathesis does not support changing the test order as of now. For an arbitrary case, it may require storing all API operations in memory which is problematic for large APIs (due to reference resolving, i.e. "methods" part could live in a different location, so it requires resolving). It could be done via a separate routine that will be called only if the order is specified explicitly. Still, I didn't find this feature important enough to increase the complexity of the codebase. I have some refactorings ongoing with all the traversing logic, so I'll re-assess it.

However, if you don't have references in path items, you can adjust the order via the `before_load_schema` hook. The trick is that Schemathesis iterates over Python dictionaries when it traverses the schema and with this hook, you can influence this order. I.e. take the value of the `paths` key and create a new dict moving path items there in the order you want:

```python
raw_schema["paths"] = {
    "/users": raw_schema["paths"]["users"],
    # ...
}
```

I think this idea can be extended to methods as well. So, with this change, Schemathesis will traverse the new dictionary and will follow the order in which you specified the keys.

But as you mentioned the operations, I see that they rely on `id`, which means they are stateful, so regular non-stateful tests will not be that effective as they don't have the information about what `id` value exists on the app side (and it will most likely lead to 404). Therefore, stateful testing via Open API links is the way to go with Schemathesis here.

**Reply by ravy** [reply]
how can this be extended to methods? do you mean i should use a tuple as the key for the raw_schema dictionary?
like ("GET", "/users"): raw_schema["paths"]["users"],

**Comment by Stranger6667** [comment]
@ravy Please, let me know if the comment above was in any way helpful. If not I'd be happy to elaborate

**Reply by ahmayun** [reply]
I have a follow up on this. Right now Schemathesis will test the paths one by one i.e. if asked to generate 10 examples per API operation it will do so like this:
```
PUT /path1 (10 requests)
GET /path1 (10 requests)
PUT /path2 (10 requests)
```

I understand from the conversation above that it is possible to adjust the order of these at the endpoint level. But what if I want to randomize the order at the request level i.e. something like this:

```
PUT /path1 (1 request)
GET /path1 (1 request)
PUT /path1 (1 request)
PUT /path2 (1 request)
GET /path1 (1 request)
... and so on...
```

Just to confirm that Schemathesis does not have this option right? Is there a technical challenge in implementing it like this? Seems to me it is a useful feature to have when the relationship between API calls is not known (via links or otherwise).

**Reply by Stranger6667** [reply]
I think you can do random requests in Python tests by using the `as_strategy` method on `schema` instances - https://github.com/schemathesis/schemathesis/blob/master/src/schemathesis/schemas.py#L499

However, specifying there is no way to specify an order for them as it is just a strategy for generating data.
